### PR TITLE
Address XDMF/Visit Issues

### PR DESCRIPTION
**Pull request by pdmullen**
Visualizing HDF5 + XDMF files generated through `develop` with VisIt produces strange artifacts wherein 2D datasets often show a protruding X3 feature (which looks like a skyscraper).  This MR attempts to fix such artifacting by adjusting the data dumped in `HyperSlab` `DataItem`s.  

Parthenon particles cannot be visualized in VisIt currently, and their corresponding mesh is not even dumped in our current XDMF files.  For the life of me, I cannot get VisIt to interpret `VXVYVZ` type geometries for `Polyvertex` meshes.  I find that in order to get Visit to read our `Polyvertex` mesh, I must recast the geometry as an array of type `{{x1, y1, z2}, {x2, y2, z2}...}`.  Even after that change, I cannot get Visit to see the `Polyvertex` mesh unless it is its own XDMF file.  Visit can still plot particles on top of field variables if both XDMF files are read and a correlation is instantiated (VisIt prompts users automatically for this).  This MR enables users to optionally dump out swarm XDMF files, which, if enabled, also dumps restructured particle geometry in the corresponding HDF5 file.  Restart files should remain unchanged.  

This is gross.  Feel free to shoot down this MR. 

## PR Checklist

- [ ] Code passes cpplint
- [ ] New features are documented.
- [ ] Adds a test for any bugs fixed. Adds tests for new features.
- [ ] Code is formatted
- [ ] Changes are summarized in CHANGELOG.md
- [ ] Change is breaking (API, behavior, ...)
  - [ ] Change is *additionally* added to CHANGELOG.md in the breaking section
  - [ ] PR is marked as breaking
  - [ ] Short summary API changes at the top of the PR (plus optionally with an automated update/fix script)
- [ ] CI has been triggered on [Darwin](https://re-git.lanl.gov/eap-oss/parthenon/-/pipelines) for performance regression tests.
- [ ] Docs build
- [ ] (@lanl.gov employees) Update copyright on changed files

### CODE REVIEW

File: src/outputs/parthenon_hdf5.cpp
Comment (Yurlungur) [inline src/outputs/parthenon_hdf5.cpp:539]: This implies (I think) that we're writing swarm positions twice, right? Since you're outputting the fused version for xdmf but still outputting the original in whatever requested shape.

Is that desirable? Or should we change swarms to always be read and written in this way? As a conceit for visualization I think I'm fine doing it this way as the original shape is better IMO and more faithful to parthenon. Could go either way.

File: src/outputs/parthenon_hdf5.cpp
Comment (Yurlungur) [inline src/outputs/parthenon_hdf5.cpp:519]: Is it necessary to fill this contiguous block like this? I feel like we could probably get around this by using HDF5's lower level memory space machinery to tell it to write swarm::x specifically into the first "column," swarm::y directly into the second column, etc. Maybe not though. Haven't thought about it carefully.

**Review by Yurlungur — COMMENTED** [review]
Hey thanks for figuring this out! And I apologize for your suffering. I don't think this MR is the optimal implementation of this machinery. But I'm not sure I care about an optimal implementation... XDMF is forcing a bizarre structure on us and I think I'm perfectly happy to just hack in a solution rather than overengineer jumping through XDMF's hoops.

> This MR attempts to fix such artifacting by adjusting the data dumped in HyperSlab DataItems.

Am I missing something? I can't find this change in the diff.

**Review by Yurlungur — APPROVED** [review]
I'm approving now to indicate my willingness to just hack this in.

File: src/outputs/parthenon_hdf5.cpp
Comment (pdmullen) [reply src/outputs/parthenon_hdf5.cpp:539]: That is correct.  This was an intentional choice; but I equally hate it.  Here was my thought process:

(1) As you mention, this is totally wasteful; I am dumping the exact same info twice; just one in a particular format to make XDMF happy.  
(2) If I just switch to the format that XDMF requires, then (a) this would be a breaking change; (b) I would subsequently have to go in and make edits to fix swarm restarts; (c) I would have to make changes to `phdf.py`.
(3) In the spirit of trying be non-breaking, I keep restarts and dumps as is, just adding this new geometry piece iff swarm XDMF files are requested (which weren't available before).

**Comment by pdmullen** [comment]
> Am I missing something? I can't find this change in the diff.

@Yurlungur  https://github.com/parthenon-hpc-lab/parthenon/pull/1079/files#diff-31b92df33d462322a957c4750de1874912d6947009746063e9684dae98314675R515

File: src/outputs/parthenon_hdf5.cpp
Comment (pdmullen) [reply src/outputs/parthenon_hdf5.cpp:519]: I was already shooting in the dark with the existing Parthenon HDF functions.  I think you are among the few that actually understands the nuts and bolts of those.  If you see how to implement what you propose straightforwardly, I would welcome that strategy with open arms.  This hack I put in is quite ugly.

File: src/outputs/parthenon_xdmf.cpp
Comment (Yurlungur) [inline src/outputs/parthenon_xdmf.cpp:515]: Oh shoot. I had to change this to support the moving coordinates machinery in 2d I think... Maybe we add a flag for that so this is `nx + (custom_coords && (nx>1))` or something?

**Review by Yurlungur — CHANGES_REQUESTED** [review]
Thanks for pointing the change out. We should test this in Assail before merging...

File: src/outputs/parthenon_hdf5.cpp
Comment (Yurlungur) [reply src/outputs/parthenon_hdf5.cpp:519]: I don't think the parthenon HDF5 functions can do it. We'd have to go to the lower-level HDF5. Let's talk either tomorrow or next week (depending on when you're around). I can try to suggest a hack.

File: src/outputs/parthenon_xdmf.cpp
Comment (pdmullen) [reply src/outputs/parthenon_xdmf.cpp:515]: I didn't think that the moving coordinates machinery even used this function.

File: src/outputs/parthenon_hdf5.cpp
Comment (Yurlungur) [reply src/outputs/parthenon_hdf5.cpp:539]: We should probably get other people's opinion on this... but I think I'm fine with your approach. Doing it the other way would require changes to both `phdf.py` and the restart reader. And I prefer the original layout (obviously... as I designed it.)

Perhaps a question is how does OpenPMD want it?

File: src/outputs/parthenon_xdmf.cpp
Comment (Yurlungur) [reply src/outputs/parthenon_xdmf.cpp:515]: Oops you're right. I must have changed this when floundering... Sorry about this and good catch.

**Review by Yurlungur — APPROVED** [review]

File: src/outputs/parthenon_hdf5.cpp
Comment (pdmullen) [reply src/outputs/parthenon_hdf5.cpp:539]: In my most recent commit, I have opted to make `write_swarm_xdmf` `false` by default.  I feel this better reflects the discussion here; i.e., if you _really_ want to read particles into VisIt, then you incur the penalty of dumping geometry twice and creating an additional XDMF file.  Furthermore, the default behavior just recovers what is the current behavior in `develop` prior to this MR (but with the field variable XDMF fix).

**Review by lroberts36 — APPROVED** [review]
I am happy to approve this if it is working for you guys, but I don't think I can provide much useful feedback since I rarely use Visit and don't know what to expect for the XDMF.

File: src/outputs/parthenon_hdf5.cpp
Comment (Yurlungur) [reply src/outputs/parthenon_hdf5.cpp:539]: I'm fine with this.

File: src/outputs/parthenon_hdf5.cpp
Comment (Yurlungur) [reply src/outputs/parthenon_hdf5.cpp:519]: We discussed offline---I think it's possible to do this with low level HDF5 but unclear if this is needed.

**Pull request merged**
